### DISCUSSION TITLE
Support for ScriptableObject Tool parameters

### DISCUSSION THREAD

**Discussion by Bodekaer** [RageAgainstThePixel/com.openai.unity#290] [Q&A] [answered]
Hi @StephenHodgson,
Do you know if `ScriptableObject` is supported as a Tool parameter?
I've tested it here and it somehow doesn't seem to include the parameters in the request.

Simple example:

```csharp
	public class ConfigSo : ScriptableObject
	{
		public string Id => _id;
		[SerializeField] [JsonProperty] [ReadOnly]
		private string _id;
        }

        public async void ToolWithScriptableObject(ConfigSo config) {
                ...
        }
```

I can always default back to using simple objects, and then just manually create the ScriptableObject afterwards. 
But wanted to check with you first :)

**Comment by Bodekaer** [comment]
REST 400 Error I get:
```json
{
  "error": {
    "message": "Invalid schema for function 'ToolWithScriptableObject': In context=(), 'required' is required to be supplied and to be an array including every key in properties. Missing 'CanRunTasksInParallel'.",
    "type": "invalid_request_error",
    "param": "tools[0].function.parameters",
    "code": "invalid_function_parameters"
  }
}
```

**Reply by Bodekaer** [reply]
This used to work before the Structured Format upgrade btw.

**Reply by StephenHodgson** [reply]
What is it currently getting serialized as?

You can check this by simply doing this:

```csharp
var schema = typeof(ScriptableObject).ToJsonSchema();
debug.log(schema);
```

**Reply by StephenHodgson** [reply]
I would also recommend that you serialize unity fields natively and use properties for json serialization.

```csharp
// Licensed under the MIT License. See LICENSE in the project root for license information.

using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using UnityEngine;

namespace OpenAI.Tests.StructuredOutput
{
    [Serializable]
    public class MathResponse
    {
        [SerializeField]
        private List<MathStep> steps;

        [JsonProperty("steps")]
        public IReadOnlyList<MathStep> Steps
        {
            get => steps;
            private set => steps = value as List<MathStep>;
        }

        [SerializeField]
        private string finalAnswer;

        [JsonProperty("final_answer")]
        public string FinalAnswer
        {
            get => finalAnswer;
            private set => finalAnswer = value;
        }
    }

    [Serializable]
    public class MathStep
    {
        [SerializeField]
        private string explanation;

        [JsonProperty("explanation")]
        public string Explanation
        {
            get => explanation;
            private set => explanation = value;
        }

        [SerializeField]
        private string output;

        [JsonProperty("output")]
        public string Output
        {
            get => output;
            private set => output = value;
        }
    }
}
```

**Reply by Bodekaer** [reply]
Ahh, I found the issue. It happens when I have class inheritance, and it doesn't include inherited properties. 
Removing `BindingFlags.DeclaredOnly` in `TypeExtensions.cs` appears to solve it.

E.g. the example below:
```csharp
[Serializable]
public class BaseClass {
	[JsonProperty("baseProperty")]
	public string BaseProperty { get; set; }
}

[Serializable]
public class MainClass : BaseClass {
	[JsonProperty("newProperty")]
	public string NewProperty { get; set; }
}
```

Serializes to:
```json
{
  "type": "object",
  "properties": {
    "newProperty": {
      "type": "string"
    }
  },
  "required": [
    "newProperty"
  ],
  "additionalProperties": false
}
```

It works fine now after removing `BindingFlags.DeclaredOnly`. 
Might cause some other side-effects though? :)

**Reply by StephenHodgson** [reply]
I see, so maybe need to tweak the binding flags for the reflection call. I was trying to prevent adding frivolous properties from some unity types.

**Reply by Bodekaer** [reply]
@StephenHodgson makes good sense. It would be nice to exclude those Unity properties. Especially on ScriptableObject :) Perhaps just only include properties with `JsonProperty()` defined?

I noticed another odd behavior that perhaps you can easily identify.
For this simple class, with an array of a child class, the property `ExcludeThis` is included in the definition, _but_ it is not included in the `required` list. This 

```csharp
		public class MainClass
		{
			public TestClass[] SomeArray { get; set; }
		}

		public class TestClass
		{
			[JsonProperty()]
			public string IncludeThis { get; set; }

			public string ExcludeThis { get; set; }
		}
```

Serializes to:
```json
{
  "type": "object",
  "properties": {
    "SomeArray": {
      "type": "array",
      "items": {
        "$ref": "#/definitions/...+TestClass"
      }
    }
  },
  "definitions": {
    "...+TestClass": {
      "type": "object",
      "properties": {
        "IncludeThis": {
          "type": "string"
        },
        "ExcludeThis": {
          "type": "string"
        }
      },
      "required": [
        "IncludeThis"
      ],
      "additionalProperties": false
    }
  },
  "required": [
    "SomeArray"
  ],
  "additionalProperties": false
}
```

Can you clarify what the best approach is here?
It would be nice if it simply excluded in the property `ExcludeThis` as it doesn't have the `JsonProperty()` attribute.

**Reply by StephenHodgson** [reply]
Well unfortunately some people want Unity Objects to be serializable and this is "a feature not a bug" kinda thing 😅 

Maybe we can offer two different modes of serialization? One that only listens to `JsonProperty` and the other that just serializes every public property with the default binding flags there are now?

Only problem is that it'll need to be documented and people will have to ***just know about it***.
I guess we could also check if the class derives from `UnityEngine.Object`. Unsure.

One other possibility is that you implement your own serializer and just pass the resulting json string in.

**Reply by Bodekaer** [reply]
Yeah I see the challenge.
I think the current setup to not include inherited properties makes sense, and then we could just add a setting/option on to include inherited properties or not. Then it's generic enough as well. 

The last example I posted above didn't use UnityObjects though. It's a separate issue I'm trying to debug.
Notice that some properties are not always included in the `required` list, and it appears to happen if the classes are listed under `"definitions"`.

It seems to work fine if I just list the same object as a variable and not an array (arrays causes it to list it under `definitions`, where some properties are then excluded from the `required` list. 
Any idea which line to tweak to always include in the `required` list. I guess all properties should always be added there.

I may just simple create a separate simpler non-Unity-inherited object for the serialization, and then cast/convert them to the Unity ScriptableObject when the GPT assistant returns them.

**Reply by StephenHodgson** [reply]
> Notice that some properties are not always included in the required list, and it appears to happen if the classes are listed under "definitions".

This is because I'm using schema `$ref` their respective requirements are listed with each class def ref

**Reply by StephenHodgson** [reply]
> I guess all properties should always be added there.

That is correct, so that there would not be a limit to the recursion and serialization depth in the json.

**Reply by StephenHodgson** [reply, marked as answer]
> I may just simple create a separate simpler non-Unity-inherited object for the serialization, and then cast/convert them to the Unity ScriptableObject when the GPT assistant returns them.

Yeah just make your own class that serializes them to the json you prefer (if you wanna use my `TypeExtensions` class as a template). Then pass the raw json schema in with your request or tool creation. (pretty sure I made it possible to define a tool via json only).

**Reply by StephenHodgson** [reply]
In the meantime I'll write up some unit tests for the nested example and do my best to sus out the right answers there along with ScriptableObject and UnityEngine types.
